batched/system.go

```diff
@@ -0,0 +1,335 @@
+package batched
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/ipfs/go-ipfs-provider/queue"
+	"strconv"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	logging "github.com/ipfs/go-log"
+	"github.com/ipfs/go-verifcid"
+	"github.com/multiformats/go-multihash"
+)
+
+var log = logging.Logger("provider.batched")
+
+type BatchProvidingSystem struct {
+	ctx   context.Context
+	close context.CancelFunc
+
+	reprovideInterval time.Duration
+	rsys              provideMany
+	keyProvider       simple.KeyChanFunc
+
+	q  *queue.Queue
+	ds datastore.Batching
+
+	provch, dynamicCh chan cid.Cid
+
+	totalProvides, lastReprovideBatchSize     int
+	avgProvideDuration, lastReprovideDuration time.Duration
+}
+
+var _ provider.System = (*BatchProvidingSystem)(nil)
+
+type provideMany interface {
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
+	Ready() bool
+}
+
+// Option defines the functional option type that can be used to configure
+// BatchProvidingSystem instances
+type Option func(system *BatchProvidingSystem) error
+
+var lastReprovideKey = datastore.NewKey("/provider/reprovide/lastreprovide")
+
+func New(provider provideMany, q *queue.Queue, opts ...Option) (*BatchProvidingSystem, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &BatchProvidingSystem{
+		ctx:   ctx,
+		close: cancel,
+
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		provch:            make(chan cid.Cid, 1),
+		dynamicCh:         make(chan cid.Cid, 1),
```

Comment:
My understanding is that having a single element buffer is the thing to do whenever you are basically streaming tasks from one goroutine to another that having a buffer doesn't hurt you much in terms of confusion in execution but may speed up processing.

For example, the "Buffered Channel 1" example in https://www.ardanlabs.com/blog/2017/10/the-behavior-of-channels.html.

I'm not attached to it here though if you think it's not worthwhile

Reply:
It might help. My main concern is that it makes writes to the channel non-blocking, which may cause a writer to think the write "succeeded" even though the reader is actually gone (shutting down).

Reply:
> which may cause a writer to think the write "succeeded" even though the reader is actually gone (shutting down)

Does this matter here? Without some feedback to the writer that the provide actually happened it has to operate on a best-effort basis anyway.

Anyhow if you just give a 👍 emoji here I'll make the channel unbuffered since I don't feel strongly.

Reply:
Logging? I don't really care much either way.

Reply:
I switched the channel to unbuffered.